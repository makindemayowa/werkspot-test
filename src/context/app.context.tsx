import * as React from 'react';
import { ITodoData } from '../types';

interface IState {
  toDos: ITodoData[];
  searchString: string
};

interface IActionCreator {
  type: string;
  data: any;
};

type IDispatch = ({ type, data }: IActionCreator) => null | void;

type IReducer = (state: IState, action: IActionCreator) => IState;

interface IActionType {
  ADD_NEW_CARD: string;
  SET_TODOS: string;
  SET_SEARCH: string;
  UPDATE_CARD_COLUMN: string;
};

const getTodos = (): ITodoData[] => {
    const todoString = localStorage.getItem('toDos');
    let todos: ITodoData[] = [];
    if(todoString) {
        todos = JSON.parse(todoString);
    };
    return todos;
}

const setTodos = (data: ITodoData[]) => localStorage.setItem('toDos', JSON.stringify(data));

const initialState: IState = {
  toDos: getTodos(),
  searchString: ''
};

export const actionType: IActionType = {
  ADD_NEW_CARD: 'ADD_NEW_CARD',
  UPDATE_CARD_COLUMN: 'UPDATE_CARD_COLUMN',
  SET_TODOS: 'SET_TODOS',
  SET_SEARCH: 'SET_SEARCH'
};

const reducer: IReducer = (state, action) => {
  switch (action.type) {
    case actionType.SET_TODOS:
      setTodos(action.data)
      return {
        ...state,
        toDos: action.data,
      };
    case actionType.UPDATE_CARD_COLUMN:
      const toDosCopy = Array.from(state.toDos);
      const cardToUpdate = toDosCopy[action.data.originalIdx].rows[action.data.cardIdx]
      toDosCopy[action.data.originalIdx].rows.splice(action.data.cardIdx, 1)
      toDosCopy[action.data.nextIdx].rows.unshift(cardToUpdate)
      setTodos(toDosCopy)
      return {
        ...state,
        toDos: toDosCopy,
      };
    case actionType.ADD_NEW_CARD:
      const cloned = Array.from(state.toDos);
      const itemIndex = action.data.columnIndex;
      const updatedColumn = cloned[itemIndex];
      updatedColumn.rows.push(action.data.newCard);
      cloned[itemIndex] = updatedColumn;
      setTodos(cloned)
      return {
        ...state,
        toDos: cloned
      };
    case actionType.SET_SEARCH:
      return {
        ...state,
        searchString: action.data
      };
    default:
      return state;
  }
};

export const AppStateContext = React.createContext<IState | undefined>(initialState);
export const AppDispatchContext = React.createContext<IDispatch | undefined>(
  undefined
);

export function useAppState() {
  const state = React.useContext(AppStateContext);

  if (state === undefined) {
    throw new Error('You can only use useAppState inside a context provider');
  }

  return state;
};

export function useAppDispatch() {
  const dispatch = React.useContext(AppDispatchContext);

  if (dispatch === undefined) {
    throw new Error(
      'You can only use useAppDispatch inside a context provider'
    );
  };

  return dispatch;
};

export function AppProvider({ children }: { children: any }) {
  const [state, dispatch] = React.useReducer(reducer, initialState);

  return (
    <AppStateContext.Provider value={state}>
      <AppDispatchContext.Provider value={dispatch}>
        {children}
      </AppDispatchContext.Provider>
    </AppStateContext.Provider>
  );
};
