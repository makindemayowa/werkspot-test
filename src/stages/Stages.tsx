import * as React from 'react';
import { ITodoData } from '../types';
import Column from './components/Column';
import './index.scss';
import { actionType, useAppDispatch, useAppState } from '../context/app.context';

export default function Stages() {
    const dispatch = useAppDispatch();
    const { toDos, searchString } = useAppState();
    const [newListValue, setNewListValue] = React.useState('');
    const [columns, setColumns] = React.useState<ITodoData[]>(toDos);

    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setNewListValue(event.target.value)
    };

    React.useEffect(() => {
        if (searchString) {
            setColumns(toDos);
            const arrToFilter = Array.from(columns);
            const updatedColumns = arrToFilter.map((arr) => {
                return ({
                    columnName: arr.columnName,
                    rows: arr.rows.filter((row) =>
                        row.title.toLowerCase().includes(searchString.toLowerCase()))
                })
            })
            setColumns(updatedColumns);
        } else {
            setColumns(toDos);
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [searchString, toDos])

    const createNewColumn = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault()
        const newColumn = {
            columnName: newListValue,
            rows: []    
        };
        const updatedColumn = [...columns, newColumn];
        setNewListValue('');
        setColumns(updatedColumn);
        return dispatch({ type: actionType.SET_TODOS, data: updatedColumn });
    }

    return (
        <div className="stages">
            {
                columns.map((column: ITodoData, idx: number) => {
                    return (
                        <Column index={idx} column={column} key={idx} />
                    )
                })
            }
            <div className="column">
                <div className="add-new-list">
                    <form onSubmit={createNewColumn}>
                        <input
                            value={newListValue}
                            onChange={onChange}
                            type="text"
                            placeholder="Add a new list...hit enter when done"
                        />
                    </form>     
                </div>
            </div>
        </div>
    );
}
