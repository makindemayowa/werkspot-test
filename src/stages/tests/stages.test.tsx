import enzyme from 'enzyme';
import renderer from 'react-test-renderer';
import { AppStateContext, AppDispatchContext } from '../../context/app.context';
import Stages from '../Stages';

const todoMocks = [
  {
    columnName: 'column-1',
    rows: [
      {
        id: 'card-1',
        title: 'Flow Meter Measurement Errors',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      },
      {
        id: 'card-2',
        title: 'Flow Meter Measurement Errors 2',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      },
    ]
  },
  {
    columnName: 'column-2',
    rows: [
      {
        id: 'card-3',
        title: 'Flow Meter Measurement Errors 3',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      },
      {
        id: 'card-4',
        title: 'Flow Meter Measurement Errors 4',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      },
    ]
  },
  {
    columnName: 'column-3',
    rows: [
      {
        id: 'card-5',
        title: 'Flow Meter Measurement Errors 5',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      },
      {
        id: 'card-6',
        title: 'Flow Meter Measurement Errors 6',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      },
      {
        id: 'card-7',
        title: 'Flow Meter Measurement Errors 7',
        description: 'https://blog.gesrepair.com/',
        tag: 'Blog Post',
        assignee: 'Peter Drury',
        dueDate: '09/15/2019'
      }
    ]
  }
];

const wrappedComponent = (Component: any) => {
  return (
    <AppStateContext.Provider value={{toDos: todoMocks, searchString: ''}}>
      <AppDispatchContext.Provider value={({ type, data }) => null}>
        <Component />
      </AppDispatchContext.Provider>
    </AppStateContext.Provider>
  )
}

test('Stages component creates cards based on context data', () => {
    const component = renderer.create(wrappedComponent(Stages));
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Should contain an input element', () => {
  const wrapper = enzyme.render(wrappedComponent(Stages));
  expect(wrapper.children().length).toEqual(4)
});