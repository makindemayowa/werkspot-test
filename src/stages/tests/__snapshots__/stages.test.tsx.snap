// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stages component creates cards based on context data 1`] = `
<div
  className="stages"
>
  <div
    className="column"
  >
    <div
      className="column-header"
    >
      <span
        className="count"
        style={
          Object {
            "backgroundColor": "#D8D8D8",
          }
        }
      >
        2
      </span>
      <span
        className="title"
      >
        column-1
      </span>
    </div>
    <div
      className="column-body"
    >
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors 2
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="add-new-card"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            onChange={[Function]}
            placeholder="Add a new card...hit enter when done"
            type="text"
            value=""
          />
        </form>
      </div>
    </div>
  </div>
  <div
    className="column"
  >
    <div
      className="column-header"
    >
      <span
        className="count"
        style={
          Object {
            "backgroundColor": "#FBEDCE",
          }
        }
      >
        2
      </span>
      <span
        className="title"
      >
        column-2
      </span>
    </div>
    <div
      className="column-body"
    >
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors 3
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors 4
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="add-new-card"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            onChange={[Function]}
            placeholder="Add a new card...hit enter when done"
            type="text"
            value=""
          />
        </form>
      </div>
    </div>
  </div>
  <div
    className="column"
  >
    <div
      className="column-header"
    >
      <span
        className="count"
        style={
          Object {
            "backgroundColor": "#90eebf",
          }
        }
      >
        3
      </span>
      <span
        className="title"
      >
        column-3
      </span>
    </div>
    <div
      className="column-body"
    >
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors 5
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors 6
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="card"
      >
        <div
          className="item-cont"
        >
          <button
            className="item"
            onClick={[Function]}
          >
            <img
              alt="Move to new column"
              src="filter-list.svg"
            />
          </button>
        </div>
        <span
          className="card-title"
        >
          Flow Meter Measurement Errors 7
        </span>
        <p
          className="description"
        >
          https://blog.gesrepair.com/
        </p>
        <div
          className="tag"
        >
          <img
            alt="category"
            src="category.svg"
          />
          <span
            className="tag-value"
          >
            Blog Post
          </span>
        </div>
        <div
          className="tag date"
        >
          <img
            alt="calendar_today"
            src="calendar.svg"
          />
          <span
            className="tag-value"
          >
            09/15/2019
          </span>
        </div>
      </div>
      <div
        className="add-new-card"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            onChange={[Function]}
            placeholder="Add a new card...hit enter when done"
            type="text"
            value=""
          />
        </form>
      </div>
    </div>
  </div>
  <div
    className="column"
  >
    <div
      className="add-new-list"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          onChange={[Function]}
          placeholder="Add a new list...hit enter when done"
          type="text"
          value=""
        />
      </form>
    </div>
  </div>
</div>
`;
